File: frontend/src/types/prediction.ts
// Prediction types

export interface HorsePrediction {
  rank: number;
  horse_id: number;
  horse_name: string;
  horse_number: number;
  score: number;
  win_probability: number;
  place_probability: number;
  popularity: number | null;
  odds: number | null;
  is_dark_horse: boolean;
  dark_horse_reason: string | null;
}

export interface PacePrediction {
  type: 'slow' | 'middle' | 'high';
  confidence: number;
  reason: string;
  advantageous_styles: string[];
  escape_count: number;
  front_count: number;
}

// 三連複（新形式: 軸2頭流し）
export interface TrioBetNew {
  type: 'pivot_2_nagashi';
  pivots: number[];
  others: number[];
  combinations: number;
  amount_per_ticket: number;
}

// 三連複（旧形式: ボックス）
export interface TrioBetOld {
  type: string;
  horses: number[];
  combinations: number;
  amount_per_ticket: number;
}

// 三連単2頭軸マルチ（新形式）
export interface TrifectaMultiBet {
  type: 'pivot_2_multi';
  pivots: number[];
  others: number[];
  combinations: number;
  amount_per_ticket: number;
}

// 両形式に対応
export interface BetRecommendation {
  // 新形式
  trio?: TrioBetNew | TrioBetOld | null;
  trifecta_multi?: TrifectaMultiBet | null;
  // 旧形式（後方互換性）
  trifecta?: any;
  exacta?: any;
  wide?: any;
  high_risk_bets?: any[];
  // 共通
  total_investment: number;
  note: string | null;
}

export interface PredictionResponse {
  race_id: number;
  model_version: string;
  predicted_at: string;
  rankings: HorsePrediction[];
  pace_prediction: PacePrediction | null;
  recommended_bets: BetRecommendation | null;
  confidence_score: number | null;
  reasoning: string | null;
}

export interface PredictionHistory {
  id: number;
  race_id: number;
  model_version: string;
  predicted_at: string;
  confidence_score: number | null;
  created_at: string;
  updated_at: string;
}

export interface PredictionHistoryListResponse {
  items: PredictionHistory[];
  total: number;
}

export interface ModelStatus {
  model_version: string;
  is_trained: boolean;
  last_trained_at: string | null;
  training_data_count: number;
  metrics: Record<string, number> | null;
}

export interface FeatureImportance {
  features: Array<{
    name: string;
    importance: number;
  }>;
}

// Pace type labels
export const PACE_LABELS: Record<PacePrediction['type'], string> = {
  slow: 'スロー',
  middle: 'ミドル',
  high: 'ハイ',
};

export const PACE_COLORS: Record<PacePrediction['type'], string> = {
  slow: '#22c55e',
  middle: '#eab308',
  high: '#ef4444',
};

File: backend/app/schemas/prediction.py
"""Prediction schemas."""

from datetime import datetime
from typing import Any

from pydantic import Field

from app.schemas.common import BaseSchema, TimestampSchema


class HorsePrediction(BaseSchema):
    """Individual horse prediction."""

    rank: int = Field(..., ge=1, description="予測順位")
    horse_id: int = Field(..., description="馬ID")
    horse_name: str = Field(..., description="馬名")
    horse_number: int = Field(..., description="馬番")
    score: float = Field(..., ge=0, le=1, description="予測スコア")
    win_probability: float = Field(..., ge=0, le=1, description="勝率予測")
    place_probability: float = Field(..., ge=0, le=1, description="複勝率予測")
    popularity: int | None = Field(None, ge=1, description="人気順")
    odds: float | None = Field(None, ge=1.0, description="単勝オッズ")
    is_dark_horse: bool = Field(False, description="穴馬フラグ")
    dark_horse_reason: str | None = Field(None, description="穴馬理由")


class PacePrediction(BaseSchema):
    """Pace prediction."""

    type: str = Field(..., description="ペースタイプ (slow/middle/high)")
    confidence: float = Field(..., ge=0, le=1, description="信頼度")
    reason: str = Field(..., description="理由")
    advantageous_styles: list[str] = Field(default_factory=list, description="有利な脚質")
    escape_count: int = Field(0, ge=0, description="逃げ馬数")
    front_count: int = Field(0, ge=0, description="先行馬数")


class BetRecommendation(BaseSchema):
    """Bet recommendation - 三連複と三連単2頭軸マルチのみ."""

    trio: dict[str, Any] | None = Field(None, description="三連複（軸2頭流し）")
    trifecta_multi: dict[str, Any] | None = Field(None, description="三連単2頭軸マルチ")
    total_investment: int = Field(0, ge=0, description="合計投資額")
    note: str | None = Field(None, description="備考")


class PredictionCreate(BaseSchema):
    """Schema for creating a prediction."""

    race_id: int = Field(..., description="レースID")


class PredictionResponse(BaseSchema):
    """Prediction response schema."""

    race_id: int
    model_version: str
    predicted_at: datetime
    rankings: list[HorsePrediction]
    pace_prediction: PacePrediction | None = None
    recommended_bets: BetRecommendation | None = None
    confidence_score: float | None = None
    reasoning: str | None = None


class PredictionHistoryResponse(TimestampSchema):
    """Prediction history response."""

    id: int
    race_id: int
    model_version: str
    predicted_at: datetime
    confidence_score: float | None = None


class PredictionHistoryListResponse(BaseSchema):
    """Prediction history list response."""

    items: list[PredictionHistoryResponse]
    total: int


class ModelStatusResponse(BaseSchema):
    """Model status response."""

    model_version: str
    is_trained: bool
    last_trained_at: datetime | None = None
    training_data_count: int = 0
    metrics: dict[str, float] | None = None


class FeatureImportanceResponse(BaseSchema):
    """Feature importance response."""

    features: list[dict[str, Any]]
